crates/fibre/src/lib.rs

```diff
@@ -14,6 +12,111 @@ cfg_if::cfg_if! {
     }
 }
 
+pub type TagId = u32;
+
+/// See SwitchReason for overall use of this type.
+#[repr(u32)]
+pub enum SwitchReasonEnum {
+    // Used to indicate that the contination has returned normally.
+    Return = 0,
+
+    // Indicates that we are suspendinga continuation due to invoking suspend.
+    // The payload is the tag to suspend with
+    Suspend = 1,
+
+    // Indicates that we are resuming a continuation via resume.
+    Resume = 2,
+}
+
+impl SwitchReasonEnum {
+    pub fn discriminant_val(&self) -> u32 {
+        // This is well-defined for an enum with repr(u32).
+        unsafe { *(self as *const SwitchReasonEnum as *const u32) }
+    }
+}
+
+/// Values of this type are passed to `wasmtime_fibre_switch` to indicate why we
+/// are switching. A nicer way of representing this type would be the following
+/// enum:
+///
+///```
+///  #[repr(C, u32)]
+///  pub enum SwitchReason {
+///      // Used to indicate that the contination has returned normally.
+///      Return = 0,
+///
+///      // Indicates that we are suspendinga continuation due to invoking suspend.
+///      // The payload is the tag to suspend with
+///      Suspend(u32) = 1,
+///
+///      // Indicates that we are resuming a continuation via resume.
+///      Resume = 2,
+///  }
+///```
+///
+/// However, we want to convert values of type `SwitchReason` to and from u64
+/// easily, which is why we need to ensure that it contains no uninitialised
+/// memory, to avoid undefined behavior.
+///
+/// We allow converting values of this type to and from u64.
+/// In that representation, bits 0 to 31 (where 0 is the LSB) contain the
+/// discriminant (as u32), while bits 32 to 63 contain the `data`.
+#[repr(C)]
+pub struct SwitchReason {
+    discriminant: SwitchReasonEnum,
+
+    // Stores tag value if `discriminant` is `suspend`, 0 otherwise.
+    data: u32,
+}
```

Comment:
Maybe add a todo about encoding this structure as a 64-bit integer?

Reply:
It's already the case that this structure is specifically designed to be encode-able as 64 bit value. See the `impl Into<u64> for SwitchDirection` further down. The structure basically just exists so that in Rust code, we don't have to fiddle with that integer directly.

Reply:
Thanks for clarifying.